### PR TITLE
Add f1_binary and f1_macro

### PR DESCRIPTION
**Pull request by huibinshen**
Change `f1` to `f1macro` to make it more explicit on what it does. `f1macro` should be used for multi-class classification datasets.
For binary classification tasks, use `f1` (with the average option 'binary').

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/sagemaker_xgboost_container/constants/xgb_constants.py [hunk 1]
```diff
@@ -16,6 +16,7 @@
     'auc',
     'aucpr',
     'f1',
+    'f1macro',
```
Comment (huibinshen) [inline src/sagemaker_xgboost_container/constants/xgb_constants.py:19]: Not completely sure about the string. Where I can see the possible metrics strings to be used for AutoPilot?

Comment (lihaife) [reply src/sagemaker_xgboost_container/constants/xgb_constants.py:19]: It is used in hyperparameter settings, e.g. eval_metric or tuning metric, in format `f1` or `validation:f1`. XGB will strip off `validation:` part.

Comment (lihaife) [reply src/sagemaker_xgboost_container/constants/xgb_constants.py:19]: Whenever this fix is accepted and merged, we should also update Autopilot pipeline recommender. In the file `algorithm/algo_metadata/config/sagemaker-xgboost/v0/sagemaker-xgboost.json`, update `F1` metric for multiclass to `validation:f1macro`. Thanks.

Comment (huibinshen) [reply src/sagemaker_xgboost_container/constants/xgb_constants.py:19]: > It is used in hyperparameter settings, e.g. eval_metric or tuning metric, in format `f1` or `validation:f1`. XGB will strip off `validation:` part.

Thanks Haifeng. In pipeline recommender (algorithm/core/const.py), the metrics are `F1` and `F1_MACRO`, is it somewhere in this package, all the metrics names are converted to lower cases such that we could use 'f1' here? If this is true, should I use `f1_macro` instead of 'f1macro'?

Comment (huibinshen) [reply src/sagemaker_xgboost_container/constants/xgb_constants.py:19]: > Whenever this fix is accepted and merged, we should also update Autopilot pipeline recommender. In the file `algorithm/algo_metadata/config/sagemaker-xgboost/v0/sagemaker-xgboost.json`, update `F1` metric for multiclass to `validation:f1macro`. Thanks.

I think we use accuracy as default metric for multi-class datasets in AutoPilot. User can override this choice to `f1macro` though.

Comment (lihaife) [reply src/sagemaker_xgboost_container/constants/xgb_constants.py:19]: The metric name conversion is defined in `algorithm/algo_metadata/config/sagemaker-xgboost/v0/sagemaker-xgboost.json`. It is up to you to define it as `f1macro` or `f1_macro`.

Note that we define multiple metrics for each problem type. We have F1 and ACCURACY defined for multiclass. ACCURACY is the default one. It doesn't mean that you cannot define other metrics for multiclass.

Comment (iyerr3) [reply src/sagemaker_xgboost_container/constants/xgb_constants.py:19]: I suggest `f1_macro` - easier to parse

Comment (huibinshen) [reply src/sagemaker_xgboost_container/constants/xgb_constants.py:19]: Alright, thanks. Will change it to `f1_macro`.

Comment (huibinshen) [reply src/sagemaker_xgboost_container/constants/xgb_constants.py]: @iyerr3 @lihaife How about XGBoost 1.0? Do we need to change it there as well?

Comment (lihaife) [reply src/sagemaker_xgboost_container/constants/xgb_constants.py]: I don't suggest back port breaking changes.

File: src/sagemaker_xgboost_container/metrics/custom_metrics.py [hunk 2]
```diff
@@ -51,6 +53,24 @@ def accuracy(preds, dtrain):
...
+    score = 0.0
+    if preds.size > 0:
+        labels = dtrain.get_label()
+        pred_labels = margin_to_class_label(preds)
+        pos_label = Counter(labels).most_common()[-1][0]  # using least frequent label as positive label
+        score = f1_score(labels, pred_labels, average='binary', pos_label=pos_label)
+    return 'f1', score
+
+
+def f1macro(preds, dtrain):
```
Comment (lihaife) [inline src/sagemaker_xgboost_container/metrics/custom_metrics.py:73]: Why not `f1_macro`? It follows the python code styling and also matches the metric name.

Comment (huibinshen) [reply src/sagemaker_xgboost_container/metrics/custom_metrics.py]: Make totally sense, will change it in the next revision.

**Comment by edwardjkim** [comment]
The CI is failing due to [docker pull rate limit](https://www.docker.com/increase-rate-limits#:~:text=To%20increase%20your%20pull%20rate,accounts%20are%20now%20in%20effect.) because it uses an anonymous user to `docker pull`. We have to ultimately fix this, but for the short term, I would simply wait a few hours and push a dummy commit, e.g., `git commit --allow-empty -m "Trigger Build"` to make the CI succeed.

**Comment by edwardjkim** [comment]
flake8 error:
```
src/sagemaker_xgboost_container/metrics/custom_metrics.py:13:1: F401 'collections.Counter' imported but unused
979
```

**Review by lihaife — APPROVED** [review]

**Comment by huibinshen** [comment]
> flake8 error:
> 
> ```
> src/sagemaker_xgboost_container/metrics/custom_metrics.py:13:1: F401 'collections.Counter' imported but unused
> 979
> ```

**Comment by huibinshen** [comment]
> flake8 error:
> 
> ```
> src/sagemaker_xgboost_container/metrics/custom_metrics.py:13:1: F401 'collections.Counter' imported but unused
> 979
> ```

Thanks, fixed.

**Comment by edwardjkim** [comment]
Sorry, another docker pull rate limit error. Is the PR ready? I can start running integration tests if ready.

**Comment by huibinshen** [comment]
> Sorry, another docker pull rate limit error. Is the PR ready other than flake8? I can start running integration tests if ready.

I can't see the failed tests, and not 100% sure if there is any other issues besides the pull rate limit error. I could trigger another dummy commit tomorrow morning to confirm.

**Comment by huibinshen** [comment]
@iyerr3 Could you have a look of this change? Thanks.

File: src/sagemaker_xgboost_container/metrics/custom_metrics.py [hunk 3]
```diff
@@ -52,6 +52,39 @@ def accuracy(preds, dtrain):
...
+
+    :param preds: Prediction values
+    :param dtrain: Training data with labels
+    :return: Metric name, f1 score
+    """
+    score = 0.0
+    if preds.size > 0:
+        labels = dtrain.get_label()
+        pred_labels = margin_to_class_label(preds)
+        # this function is used only for AutoPilot and the least frequent label is already encoded as 1.
```
Comment (iyerr3) [inline src/sagemaker_xgboost_container/metrics/custom_metrics.py:81]: There's nothing that stops external customers from using this directly, so I wouldn't make this assumption. 
Also I'm not sure what the implications of the 'least frequent label is already encoded as 1' is. Can you restate the assumption here to ensure behavior is clear. 

Also do we need validation for the number of labels to be `2`?

Comment (lihaife) [reply src/sagemaker_xgboost_container/metrics/custom_metrics.py:81]: Why does this assumption matter? For F1, we just care what's positive/negative.

Comment (huibinshen) [reply src/sagemaker_xgboost_container/metrics/custom_metrics.py:81]: To make f1 binary work, the only requirement is the label is encoded in {0, 1}. Taking least frequent label as 1 is only a behavior AutoPilot enforces. The customer could also use more frequent label as 1.

A question here is that are the labels already encoded in {0, 1}? If not, we need to have another label encoder to make f1 binary works.

Comment (lihaife) [reply src/sagemaker_xgboost_container/metrics/custom_metrics.py:81]: @huibinshen Note that you get the labels from `margin_to_class_label` function, which always encodes in `{0,1}`. So your code is fine. More importantly, XGBoost is a product itself. You shouldn't add comments about Autopilot's behavior here. Just remove the comment, which is the root cause of confusions.

Comment (huibinshen) [reply src/sagemaker_xgboost_container/metrics/custom_metrics.py:81]: Thanks Haifeng, will remove the comment in the next revision.

@iyerr3 Do you have other comments? I hope the explanation above is clear.

Comment (iyerr3) [reply src/sagemaker_xgboost_container/metrics/custom_metrics.py]: Thanks for removing the comment. 
My question on validation still stands: if customer uses `f1_binary` with multi-class, then they should get a Customer Error. Is that validation happening somewhere that I missed?

Comment (huibinshen) [reply src/sagemaker_xgboost_container/metrics/custom_metrics.py]: Hi Rahul, I added a customer error in the new revision.

File: src/sagemaker_xgboost_container/constants/xgb_constants.py [hunk 4]
```diff
@@ -42,6 +44,7 @@
 
 LOGISTIC_REGRESSION_LABEL_RANGE_ERROR = "label must be in [0,1] for logistic regression"
 MULTI_CLASS_LABEL_RANGE_ERROR = "label must be in [0, num_class)"
+MULTI_CLASS_F1_BINARY_ERROR = "Target is multiclass but average='binary'"
```
Comment (lihaife) [inline src/sagemaker_xgboost_container/constants/xgb_constants.py:47]: Why do we throw this error?

Comment (huibinshen) [reply src/sagemaker_xgboost_container/constants/xgb_constants.py:47]: When user uses F1 binary score for multiclass dataset, f1_sccore(..., average="binary") will throw this error. This is because a mis-configuration, so it is now thrown as a CustomerError.

**Review by edwardjkim — APPROVED** [review]
Changes including customer error look good to me. Integration tests all succeeded.
